File: src/lib/ipc/fhirPackages.ts
// FHIR profile packages — IPC bindings.
//
// A package is a FHIR NPM `.tgz` (hl7.fhir.r4.core, a national IG, a site's
// own profiles). Installing one distils its StructureDefinitions into
// <config>/BridgeLab/fhir-packages/ so startup does not re-read the archive.

import { invoke } from '@tauri-apps/api/core';

export interface FhirPackageInfo {
	name: string;
	version: string;
	title: string;
	fhir_version: string;
	profile_count: number;
}

export async function listFhirPackages(): Promise<FhirPackageInfo[]> {
	return invoke('fhir_packages_list');
}

export async function reloadFhirPackages(): Promise<FhirPackageInfo[]> {
	return invoke('fhir_packages_reload');
}

/** Install from a `.tgz` on disk. Slow for a core package — show progress. */
export async function installFhirPackage(path: string): Promise<FhirPackageInfo[]> {
	return invoke('fhir_packages_install', { path });
}

export async function removeFhirPackage(name: string, version: string): Promise<FhirPackageInfo[]> {
	return invoke('fhir_packages_remove', { name, version });
}

export async function fhirPackagesDir(): Promise<string> {
	return invoke('fhir_packages_dir');
}

File: src-tauri/src/commands/fhir_packages.rs
//! IPC commands for the FHIR profile package manager.

use std::path::PathBuf;

use tauri::State;

use crate::licensing::feature_gate;
use crate::parser::fhir::profile::{package, PackageInfo, ProfileRegistry};

/// Packages currently loaded.
#[tauri::command]
pub fn fhir_packages_list(registry: State<'_, ProfileRegistry>) -> Vec<PackageInfo> {
    registry.packages()
}

/// Re-read the installed packages from disk.
#[tauri::command]
pub fn fhir_packages_reload(registry: State<'_, ProfileRegistry>) -> Result<Vec<PackageInfo>, String> {
    registry.reload()?;
    Ok(registry.packages())
}

/// Install a `.tgz` the user picked, then reload.
///
/// Reading the archive is the slow part (a core package holds thousands of
/// files), so it happens once here and the distilled index is what the app
/// loads from then on.
#[tauri::command]
pub fn fhir_packages_install(
    path: String,
    registry: State<'_, ProfileRegistry>,
) -> Result<Vec<PackageInfo>, String> {
    feature_gate::require("fhir_profile_validation")?;
    package::install(&PathBuf::from(path))?;
    registry.reload()?;
    Ok(registry.packages())
}

#[tauri::command]
pub fn fhir_packages_remove(
    name: String,
    version: String,
    registry: State<'_, ProfileRegistry>,
) -> Result<Vec<PackageInfo>, String> {
    package::remove(&name, &version)?;
    registry.reload()?;
    Ok(registry.packages())
}

/// Directory the packages live in, so the UI can offer to open it.
#[tauri::command]
pub fn fhir_packages_dir() -> Result<String, String> {
    package::packages_root()
        .map(|p| p.display().to_string())
        .ok_or_else(|| "Could not determine the config directory".into())
}
